libs/core/type_support/include/hpx/type_support/relocate_at.hpp

```diff
@@ -0,0 +1,151 @@
+//  Copyright (c) 2023 Isidoros Tsaousis-Seiras
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#pragma once
+
+#include <hpx/concepts/concepts.hpp>
+#include <hpx/type_support/construct_at.hpp>
+#include <hpx/type_support/is_relocatable.hpp>
+#include <hpx/type_support/is_trivially_relocatable.hpp>
+
+#include <cstring>
+#include <type_traits>
+
+#if defined(HPX_HAVE_P1144_STD_RELOCATE_AT)
+#include <memory>
+#endif
+
+namespace hpx::detail {
+#if __cplusplus <= 201703L
+    // pre c++17 std::destroy_at can be used only on non-array types
+    template <typename T,
+        HPX_CONCEPT_REQUIRES_(std::is_destructible_v<T> && !std::is_array_v<T>)>
+#else
+    // c++17 std::destroy_at can be used on array types, destructing each element
+    template <typename T, HPX_CONCEPT_REQUIRES_(std::is_destructible_v<T>)>
+#endif
```

Comment:
I'm confused by this. You're unconditionally using `destroy_guard` inside `relocate`, and unconditionally using `destroy_at` inside `destroy_guard`... so don't these comments simply indicate that you have a bug in C++17 when `relocate` is used on array types? ...Ah, except that `relocate` _already_ can't be used on arrays, because `is_constructible_v<int[4], int[4]>` is false. So all this ifdef stuff is simply unnecessary. You can replace lines 22–29 with simply
```
    template<typename T>
```

Reply:
The idea was to make `destroy_guard` usable outside relocation too. The checks done here are not to prevent invalid relocation, but invalid use of this facillity.

Reply:
Okay. Personally I'd still say ["You Ain't Gonna Need It"](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it); and if you think you _will_ need a bulletproof reusable `destroy_guard` then I'd say you should be investing _more_ in it — it should be in its own file instead of `relocate_at.hpp`, and named `hpx::destroy_guard` instead of `hpx::detail::destroy_guard`, and you should make it _work_ for arrays in C++17 instead of giving it different behavior depending on the C++ version. But that all sounds tangential to my interest in this PR, at least.

Reply:
Yea, I do agree with that. We are almost certainly not going to use this elsewhere. 

I do however have to leave the `HPX_CONCEPT_REQUIRES` etc there for the address sanitizer's warnings. 

Thanks for your interest in enforcing best practices, I appreciate it :)

Reply:
> I do however have to leave the `HPX_CONCEPT_REQUIRES` etc there for the address sanitizer's warnings.

*quizzical look* I don't understand why a requires-clause/enable_if/etc would have anything to do with address sanitizer. Or even (assuming you misspoke) what it would have to do with static analysis. This is again a tangent and non-blocking, but I'm curious if you could paste the warning/error that you get by omitting the `HPX_CONCEPT_REQUIRES`?

Reply:
It was a verbal warning to me by a colleague :P . I will see if I can find out more and get back to you!

Reply:
@Quuxplusone Doesn't it making it safer constraining the types that can be destroyed/guarded from the destroy guard? Def not a strong opinion of mine...